### DISCUSSION TITLE
sumtype array initialization

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#14604] [Ideas]
**V version:** latest
**OS:** ubuntu

**What did you do?**
```v
type Abc = int | string

fn main() {
	mut abc := []Abc{init: [1, 2, 'test']}
}
```

**What did you expect to see?**
either a syntax error or that it works.
 
**What did you see instead?**
```
/home/daniel/.cache/v/main.tmp.c: In function ‘main’:
/home/daniel/.cache/v/main.tmp.c:9560:69: error: declaration of type name as array of voids
 9560 |  Abc _val_49 = new_array_from_c_array(3, 3, sizeof(void), _MOV((void[3]){
      |                                                               
```

**Comment by danieldaeschle** [comment]
new error:
```
main.v:4:32: error: invalid array element: expected `int`, not `string`
    2 | 
    3 | fn main() {
    4 |     mut abc := []Abc{init: [1, 2, 'test']}
      |                                   ~~~~~~
    5 | }
```

**Comment by mcastorina** [comment]
That new error is coming from arrays needing to have homogenous element types. Casting the first element to `Abc` compiles correctly:

```v
type Abc = int | string

fn main() {
    mut abc := [Abc(1), 2, 'test']
}
```

I think this issue can be closed.

**Comment by remimimimimi** [comment]
I agree with @mcastorina, issue closed

**Comment by danieldaeschle** [comment]
This is not the solution to the problem.

**Comment by Terisback** [comment]
ok, do this then
```v
type Abc = int | string

fn main() {
	mut abc := []Abc{init: [Abc(1), 2, 'test']}
}
```

**Comment by danieldaeschle** [comment]
I know this works, but to me, my example should work as well.

**Comment by ArtemkaKun** [comment]
Hi @danieldaeschle 

2 problems with this issue:
1. Idk if this worked in the past, but code construction `[]<type>{init: [<val>, <val>, <val>']}` cannot be used for any type of array. _init:_ can only have a single value, not an array of values. See [arrays](https://github.com/vlang/v/blob/master/doc/docs.md#arrays) documentation
2. As @mcastorina mentioned, arrays in V are homogeneous and it can't understand what do you want in the case of sumtype. That's why you need to specify a type (`abc := [Abc(1), 2, 'test']`)

Maybe this will be better to close the issue since a lot of stuff in V changed and open a new one with e.g. **feature request** instead of a **bug** label, what do you think?

**Comment by tobyjwebb** [comment]
I would consider it a feature request also.

The syntax that was initially proposed would conflict with the how `init:` works currently, so I'd say that's not a good idea:
```v
mut abc := []Abc{init: [1, 2, 'test']}
```

Some type of alternative syntax could be used (a different keyword than `init`):
```v
mut abc := []Abc{[1, 2, 'test']}            // ? 1
mut def := []Abc{from: [1, 2, 'test']}      // ? 2
mut def := []Abc{init_from: [1, 2, 'test']} // ? 3
```

If I were to implement this, I would probably chose option `2`. Thoughts?

**Comment by mcastorina** [comment]
I'm a bit confused on what this new feature will do.
